libqtile/core/manager.py

```diff
@@ -174,9 +174,38 @@ def load_config(self, initial: bool = False) -> None:
         if self.config.reconfigure_screens:
             hook.subscribe.screen_change(self.cmd_reconfigure_screens)
 
+        # If user wants wake_up hooks we need to add a dbus rule
+        if "wake_up" in hook.subscriptions:
+            task = asyncio.create_task(
+                add_signal_receiver(
+                    self.on_wake,
+                    session_bus=False,
+                    dbus_interface="org.freedesktop.login1.Manager",
+                    bus_name="org.freedesktop.login1",
+                    check_service=True,
+                )
+            )
+            task.add_done_callback(self._wake_callback)
+
         if initial:
             hook.fire("startup_complete")
 
+    def _wake_callback(self, task: asyncio.Task) -> None:
+        if not task.result():
+            logger.warning("Unable to subscribe to wake events")
+
+    # We need to ignore typing here. msg is a dbus_next.Message object but dbus_next may not be installed
+    def on_wake(self, msg):  # type: ignore
+        # We only want the PrepareForSleep or PrepareForSuspend signals
+        if not msg.member.startswith("PrepareFor"):
+            return
+
+        sleeping = msg.body[0]
+
+        # Value is False when the event is over i.e. the machine has woken up
+        if not sleeping:
+            hook.fire("wake_up")
```

Comment:
There isn't much use of `self` here. Could you move this all into `utils`?

Reply:
Yes - that would be neater. Let me redo it.